tests/testthat/test-add_difference_row.tbl_summary.R

```diff
@@ -0,0 +1,140 @@
+    trial |>
+      dplyr::filter(grade != "II") |>
+      cardx::ard_stats_t_test(variables = "age", by = "grade") |>
+      dplyr::select(-"fmt_fn")
+  )
+  # Response I vs III
+  expect_equal(
+    tbl$cards$add_difference_row$response$`'I' vs. 'III'` |>
+      dplyr::select(-c("stat_fmt", "fmt_fn")),
+    trial |>
+      dplyr::filter(grade != "II") |>
+      cardx::ard_stats_prop_test(variables = "response", by = "grade") |>
+      dplyr::select(-"fmt_fn")
+  )
+})
+
+test_that("add_difference_row(reference) messaging", {
+  # bad input to `reference` arg
+  expect_snapshot(
+    error = TRUE,
+    trial |>
+      tbl_summary(by = grade, include = c(age, response), missing = "no") |>
+      add_difference_row(reference = "XXX")
+  )
+})
+
+test_that("add_difference_row() messaging", {
+  expect_snapshot(
+    error = TRUE,
+    tbl_summary(trial, include = age) |>
+      add_difference_row(reference = "I")
+  )
+
+  expect_snapshot(
+    tbl <- tbl_summary(trial, by = grade, include = response, percent = "row") |>
+      add_difference_row(reference = "I")
+  )
+})
+
+test_that("add_difference_row(test)", {
+  # works with a custom test
+  expect_equal(
+    trial |>
+      tbl_summary(by = grade, include = age, missing = "no") |>
+      add_difference_row(
+        reference = "I",
+        statistic = everything() ~ c("{estimate}", "{conf.low}, {conf.high}", "{p.value}"),
+        test = age ~ \(data, variable, by, ...) t.test(reformulate(by, variable), data = data) |> broom::tidy()
+      ) |>
+      getElement("cards") |>
+      getElement("add_difference_row") |>
+      getElement("age") |>
+      getElement("'I' vs. 'III'") |>
+      dplyr::select(-c("stat_fmt", "fmt_fn", "context", "stat_label")) |>
+      dplyr::filter(stat_name %in% c("estimate", "conf.low", "conf.high", "p.value")),
+    trial |>
+      dplyr::filter(grade != "II") |>
+      cardx::ard_stats_t_test(variables = "age", by = "grade") |>
+      dplyr::select(-c("fmt_fn", "context", "stat_label")) |>
+      dplyr::filter(stat_name %in% c("estimate", "conf.low", "conf.high", "p.value")),
+    ignore_attr = TRUE
+  )
+})
+
+test_that("add_difference_row(test) messaging", {
+  # errors when the test return is in the wrong format
+  expect_snapshot(
+    error = TRUE,
+    tbl_summary(trial, by = trt, include = age) |>
+      add_difference_row(reference = "Drug A", test = age ~ \(...) letters)
+  )
+
+  # runs ok when test errors
+  expect_snapshot(
+    tbl <- tbl_summary(trial, by = trt, include = age) |>
+      add_difference_row(reference = "Drug A", test = age ~ \(...) stop("oy!"))
+  )
+})
+
+test_that("add_difference_row.tbl_summary(group)", {
+  trial_group <- trial |>
+    dplyr::mutate(.by = trt, id = dplyr::row_number())
+
+  tbl_groups <-
+    trial_group |>
+    select(trt, id, stage, marker, age) %>%
+    tbl_summary(
+      by = trt,
+      missing = "no",
+      include = "age"
+    ) |>
+    add_difference_row(
+      test = list(age = "ancova_lme4"),
```

Comment:
Ok I see! I think still the error messaging can guide noobs like me towards this a bit more ;)

Reply:
What do you have in mind? This code runs without an error, right?

Reply:
yes, sorry I was still talking about the previous case, where I added directly the function wrongly